File: newScript.rb
#!/usr/bin/env ruby

path = ARGV[0]
fail "specify filename to create" unless path

path_list = path.split('/')
dir = path_list[0..-2].join('/')

# sample script
File.open(path, "w") do |f|
	f.puts "#!/usr/bin/env ruby -w"
	f.puts "\n# Definition of a sample user class"
	f.puts "class User"
	f.puts "\n    # Name of user"
	f.puts "    attr_accessor :name"
	f.puts "\n    # What does the user say"
	f.puts "    def speak"
	f.puts "        return 'HelloWorld!'"
	f.puts "    end"
	f.puts "\nend #class User"
	f.puts "\n# Only run the following code when this file is the main file being run"
	f.puts "if __FILE__ == $0"
	f.puts "end #if"
end

File.chmod(0755, path)
system "open", path

# unit test script
test_dir = dir+"/test"
Dir.mkdir test_dir unless Dir.exist? test_dir

test_path = test_dir+"/tc_"+path_list[-1]
File.open(test_path, "w") do |f|
	f.puts "require 'test/unit'"
	f.puts "require_relative \'../#{path_list[-1]}\'"
	f.puts "\nclass UserTest < Test::Unit::TestCase"
	f.puts "    def test_speak"
	f.puts "        u = User.new"
	f.puts "        assert_equal 'HelloWorld!', u.speak"
	f.puts "    end #test_speak"
	f.puts "end #class"
end

# Rakefile
rake_path = dir+"/Rakefile"
File.open(rake_path, "w") do |f|
	f.puts "require 'rdoc/task'"
	f.puts "require 'rake/testtask'"
	f.puts "\nRake::RDocTask.new do |t|"
	f.puts "    t.rdoc_files.include '*.rb'"
	f.puts "    t.options << '--diagram'"
	f.puts "end"
	f.puts "\nRake::TestTask.new do |t|"
	f.puts "    t.test_files = FileList['test/tc_*.rb']"
	f.puts "end"
end

File: newScript.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

' a script to create new python project '

__author__ = "Bo DONG"

import sys
import os
import argparse


def main():
    description = "This is a script."
    usage = "python3 THIS_SCRIPT [options]"
    parser = argparse.ArgumentParser(
        description=description,
        usage=usage)
    parser.add_argument(
        '-d', '--dir',
        action="store",
        help="target dir", 
        default=None)
    args = parser.parse_args()

    print("Arguments:", args)
    print("Target dir:", args.dir)

    path = args.dir
    if os.path.isdir(path):
        print("Dir '{}' already exists. Please specify a new dir".format(path))
    else:
        # get project name
        name = path.split('/')[-1]

        # ****************
        # create dirs
        # ****************
        print("Creating dir: {}".format(path))
        os.makedirs(path+"/"+name)
        os.makedirs(path+"/tests")
        os.makedirs(path+"/docs")

        # ****************
        # create files
        # ****************
        header = """#!/usr/bin/env python3
# -*- coding: utf-8 -*-

' a script to do work '
__author__ = "Bo DONG"

import sys
import os
"""
        footer = """
# only run the following code when this file is the main file being run
if __name__=='__main__':
    main()

"""

        # create script
        body = """
import argparse


def func():
    return "Hello World!"

def main():
    description = "This is a script."
    usage = "python3 THIS_SCRIPT [options]"
    parser = argparse.ArgumentParser(
        description=description,
        usage=usage)
    parser.add_argument(
        '-d', '--dir',
        action="store",
        help="target dir", 
        default=None)
    args = parser.parse_args()

    print("Arguments:", args)
    print("Target dir:", args.dir)

    print(func())

"""
        with open(path+"/"+name+"/"+name+".py", "w") as f:
            f.write(header)
            f.write(body)
            f.write(footer)

        # create context
        context_body = """
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '{}')))

import {}
""".format(name, name)
        with open(path+"/tests/context.py", "w") as f:
            f.write(header)
            f.write(context_body)

        # create tests
        test_body = """import unittest

from context import {}

class Test_func(unittest.TestCase):
    def test_func(self):
        self.assertEqual({}.func(), "Hello World!")
""".format(name, name)

        with open(path+"/tests/test_"+name+".py", "w") as f:
            f.write(header)
            f.write(test_body)

        # create Makefile
        make_body = """
init:
    pip install -U pytest

test:
    pytest

PHONY: init test
"""
        with open(path+"/Makefile", "w") as f:
            f.write(make_body)

        # create README and docs
        readme_body = """
[TOC]

# Introduction

This script can do work

"""
        with open(path+"/README.md", "w") as f:
            f.write(readme_body)

        with open(path+"/docs/"+name+".md", "w") as f:
            f.write(readme_body)

        # create .gitignore
        gitignore_body = """
__pycache__/
*.py[cod]
*$py.class
"""
        with open(path+"/.gitignore", "w") as f:
            f.write(gitignore_body)


# only run the following code when this file is the main file being run
if __name__ == '__main__':
    main()
